usr/src/uts/common/io/usb/clients/ccid/ccid.c

```diff
@@ -0,0 +1,4426 @@
+	    CCID_REQUEST_GET_PARAMS, 0, 0, 0, &cc)) != 0) {
+		return (ret);
+	}
+
+	if ((ret = ccid_command_queue(ccid, cc)) != 0)
+		goto done;
+
+	ccid_command_poll(ccid, cc);
+
+	if (cc->cc_state != CCID_COMMAND_COMPLETE) {
+		ret = EIO;
+		goto done;
+	}
+
+	ccid_command_status_decode(cc, &crs, &cis, NULL);
+	if (crs != CCID_REPLY_STATUS_COMPLETE) {
+		if (cis == CCID_REPLY_ICC_MISSING) {
+			ret = ENXIO;
+		} else {
+			ret = EIO;
+		}
+		goto done;
+	}
+
+	/*
+	 * The protocol is in ch_param2 of the header.
+	 */
+	prot = ccid_command_resp_param2(cc);
+	mlen = ccid_command_resp_length(cc);
+	cpbuf = cc->cc_response->b_rptr + sizeof (ccid_header_t);
+
+	ret = 0;
+	switch (prot) {
+	case 0:
+		if (mlen < sizeof (ccid_params_t0_t)) {
+			ret = EOVERFLOW;
+			goto done;
+		}
+		*protp = ATR_P_T0;
+		bcopy(cpbuf, &paramsp->ccp_t0, sizeof (ccid_params_t0_t));
+		break;
+	case 1:
+		if (mlen < sizeof (ccid_params_t1_t)) {
+			ret = EOVERFLOW;
+			goto done;
+		}
+		*protp = ATR_P_T1;
+		bcopy(cpbuf, &paramsp->ccp_t1, sizeof (ccid_params_t1_t));
+		break;
+	default:
+		ret = ECHRNG;
+		break;
+	}
+
+done:
+	ccid_command_free(cc);
+	return (ret);
+}
+
+static void
+ccid_hw_error(ccid_t *ccid, ccid_intr_hwerr_t *hwerr)
+{
+	ccid_slot_t *slot;
+
+	/* Make sure the slot number is within range. */
+	if (hwerr->cih_slot >= ccid->ccid_nslots)
+		return;
+
+	slot = &ccid->ccid_slots[hwerr->cih_slot];
+
+	/* The only error condition defined by the spec is overcurrent. */
+	if (hwerr->cih_code != CCID_INTR_HWERR_OVERCURRENT)
+		return;
+
+	/*
+	 * The worker thread will take care of this situation.
+	 */
+	slot->cs_flags |= CCID_SLOT_F_INTR_OVERCURRENT;
+	ccid_worker_request(ccid);
+}
+
+static void
+ccid_intr_pipe_cb(usb_pipe_handle_t ph, usb_intr_req_t *uirp)
+{
+	mblk_t *mp;
+	size_t msglen, explen;
+	uint_t i;
+	boolean_t change;
+	ccid_intr_hwerr_t ccid_hwerr;
+	ccid_t *ccid = (ccid_t *)uirp->intr_client_private;
+
+	mp = uirp->intr_data;
+	if (mp == NULL)
+		goto done;
+
+	msglen = msgsize(mp);
+	if (msglen == 0)
+		goto done;
+
+	switch (mp->b_rptr[0]) {
+	case CCID_INTR_CODE_SLOT_CHANGE:
+		mutex_enter(&ccid->ccid_mutex);
+		ccid->ccid_stats.cst_intr_slot_change++;
+
+		explen = 1 + ((2 * ccid->ccid_nslots + (NBBY-1)) / NBBY);
+		if (msglen < explen) {
+			ccid->ccid_stats.cst_intr_inval++;
+			mutex_exit(&ccid->ccid_mutex);
+			goto done;
+		}
+
+		change = B_FALSE;
+		for (i = 0; i < ccid->ccid_nslots; i++) {
+			uint_t byte = (i * 2 / NBBY) + 1;
+			uint_t shift = i * 2 % NBBY;
+			uint_t present = 1 << shift;
+			uint_t delta = 2 << shift;
+
+			if (mp->b_rptr[byte] & delta) {
+				ccid_slot_t *slot = &ccid->ccid_slots[i];
+
+				slot->cs_flags &= ~CCID_SLOT_F_INTR_MASK;
+				slot->cs_flags |= CCID_SLOT_F_CHANGED;
+				if (mp->b_rptr[byte] & present) {
+					slot->cs_flags |= CCID_SLOT_F_INTR_ADD;
+				} else {
+					slot->cs_flags |= CCID_SLOT_F_INTR_GONE;
+				}
+				change = B_TRUE;
+			}
+		}
+
+		if (change) {
+			ccid_worker_request(ccid);
+		
```

Comment:
I'm not sure EBUSY would make sense. If a user doesn't have an exclusive hold, they're not allowed to poll just as they're not allowed to read or write. The chpoll(9e) manpage doesn't mention any restrictions on the errno we are allowed to return, and poll(2) does list a few but I'm not sure that list is exhaustive. The only other errno that would possibly make sense if we were to restrict ourselves to what poll(2) states would be EINVAL, as the minor is clearly in an invalid state for polling when not held exclusively. But that doesn't match at all what poll(2) states about the reasons for EINVAL.

Reply:
I disagree. Another user has an exclusive hold and is presumably
therefore using the slot. When it's done it will become available again.
EBUSY is exactly the right return here. There is no permissions involved
here which is what EACCESS is for

Reply:
All this checks is whether the minor which is polled has an exclusive hold. If it doesn't that doesn't mean anything about whether another minor may have an exclusive hold or not, so EBUSY just isn't right as we just don't know whether it's busy.

Perhaps EINVAL is the better error code here, as a minor that hasn't a exclusive hold is in an invalid state for polling. Or perhaps this should just return 0 and indicate neither POLLIN nor POLLOUT as there's just nothing going on currently?

Reply:
I had the test inverted in my head, so you're right, EBUSY isn't correct; thanks. I think EINVAL is the best option here.

Reply:
Thinking more about this over the past day, I decided to return no error and just return no events if not in a transaction. The main reason is that with no transaction you can't write or read, and thats what no events indicate. Also EINVAL is documented in poll(2) to indicate a handful of special cases that don't apply here.

Reply:
OK, but please make sure this is documented clearly.